### PR TITLE
refactor: passing QueryContext to RegionServer

### PR DESCRIPTION
**Pull request by Kelvinyu1117**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## Refer to a related PR or issue link (optional)
close https://github.com/GreptimeTeam/greptimedb/issues/3752

## What's changed and what's your intention?

Allow passing the query context to the region server.

Please explain IN DETAIL what the changes are in this PR and why they are needed:

- Summarize your change (**mandatory**)
- How does this PR work? Need a brief introduction for the changed logic (optional)
- Describe clearly one logical change and avoid lazy messages (optional)
- Describe any limitations of the current code (optional)

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.
- [x]  This PR does not require documentation updates.

### CODE REVIEW

File: src/session/src/context.rs [hunk 1]
```diff
@@ -84,15 +84,25 @@ impl Clone for QueryContext {
...
+            },
+            None => QueryContext {
+                current_catalog: DEFAULT_CATALOG_NAME.to_string(),
+                current_schema: DEFAULT_SCHEMA_NAME.to_string(),
+                current_user: ArcSwap::new(Arc::new(None)),
+                timezone: ArcSwap::new(Arc::new(get_timezone(None).clone())),
+                sql_dialect: Arc::new(GreptimeDbDialect {}),
+                extension: HashMap::new(),
+                configuration_parameter: Default::default(),
+            },
```
Comment (MichaelScofield) [inline src/session/src/context.rs:105]: Try this:

```suggestion
        let mut builder = QueryContextBuilder::default();
        if let Some(ctx) = &value.query_context {
            builder = builder
                .current_catalog(ctx.current_catalog.clone())
                .current_schema(ctx.current_schema.clone())
                .timezone(ArcSwap::new(Arc::new(parse_timezone(Some(&ctx.timezone)))))
                .extension(ctx.extension.clone());
        }
        builder.build()
```

File: src/session/src/context.rs [hunk 2]
```diff
@@ -169,6 +179,10 @@ impl QueryContext {
         self.extension.get(key.as_ref()).map(|v| v.as_str())
     }
 
+    pub fn to_extension(&self) -> HashMap<String, String> {
```
Comment (MichaelScofield) [inline src/session/src/context.rs:182]:
```suggestion
    pub fn extensions(&self) -> HashMap<String, String> {
```

File: src/session/src/context.rs (hunk 1 above)
Comment (Kelvinyu1117) [reply src/session/src/context.rs:105]: `build()` return `QueryContextRef`, should I change the interface?

Comment (evenyag) [reply src/session/src/context.rs:105]: Sure.

File: src/session/src/context.rs (hunk 2 above)
Comment (evenyag) [reply src/session/src/context.rs:182]: Maybe `extension()`? As the name of the field is `extension`.

It is better to rename the field to `extensions`? @MichaelScofield

Comment (Kelvinyu1117) [reply src/session/src/context.rs:182]: I think you can't, because Rust dont support function overloading.

Comment (MichaelScofield) [reply src/session/src/context.rs:182]: @evenyag me ok

Comment (Kelvinyu1117) [reply src/session/src/context.rs]: fixed.

File: src/session/src/context.rs (hunk 1 above)
Comment (Kelvinyu1117) [reply src/session/src/context.rs]: fixed.

File: src/query/src/dist_plan/merge_scan.rs [hunk 3]
```diff
@@ -192,7 +195,12 @@ impl MergeScanExec {
                     header: Some(RegionRequestHeader {
                         tracing_context: tracing_context.to_w3c(),
                         dbname: dbname.clone(),
-                        timezone: tz.clone(),
+                        query_context: Some(QueryContext {
+                            current_catalog: current_catalog.clone(),
+                            current_schema: current_schema.clone(),
+                            timezone: timezone.clone(),
+                            extension: extensions.clone(),
```
Comment (evenyag) [inline src/query/src/dist_plan/merge_scan.rs:202]: Could you please rename the proto message field to extensions?
https://github.com/GreptimeTeam/greptime-proto/pull/156/files#r1587153168

File: Cargo.toml [hunk 4]
```diff
@@ -115,7 +115,7 @@ etcd-client = { git = "https://github.com/MichaelScofield/etcd-client.git", rev
 fst = "0.4.7"
 futures = "0.3"
 futures-util = "0.3"
-greptime-proto = { git = "https://github.com/GreptimeTeam/greptime-proto.git", rev = "2c14c6e22dfe957f40bb88dd01fb8530656de89b" }
+greptime-proto = { git = "https://github.com/Kelvinyu1117/greptime-proto", branch = "refactor/kelvin-add-query-context" }
```
Comment (evenyag) [inline Cargo.toml:118]: Just a remarker for updating the `greptime-proto`

**Review by evenyag — APPROVED** [review]
Mostly LGTM.

File: src/query/src/dist_plan/merge_scan.rs (hunk 3 above)
Comment (Kelvinyu1117) [reply src/query/src/dist_plan/merge_scan.rs:202]: done

File: Cargo.toml (hunk 4 above)
Comment (Kelvinyu1117) [reply Cargo.toml:118]: I believe you may want to merge the proto file first?

Comment (evenyag) [reply Cargo.toml:118]: We could merge https://github.com/GreptimeTeam/greptime-proto/pull/156 once this PR gets two approvals.

Comment (MichaelScofield) [reply Cargo.toml:118]: @Kelvinyu1117 proto is merged

**Comment by fengjiachun** [comment]
Hi @Kelvinyu1117 , there are some conflicts need to be resolved.

**Comment by killme2008** [comment]
Good job! Please resolve the conflicts, then we can merge the PR.  @Kelvinyu1117

File: Cargo.toml (hunk 4 above)
Comment (Kelvinyu1117) [reply Cargo.toml]: updated proto's repo.

**Comment by Kelvinyu1117** [comment]
> Hi @Kelvinyu1117 , there are some conflicts need to be resolved.

Should be fixed.

**Comment by WenyXu** [comment]
> > Hi @Kelvinyu1117 , there are some conflicts need to be resolved.
> 
> Should be fixed.

Hi @Kelvinyu1117 , The coverage tests failed(`test_do_auth_with_user_provider` and `test_do_auth_without_user_provider`). We need to wrap the `query_context` with `Arc` in the `do_auth` method.

File: src/query/src/dist_plan/merge_scan.rs [hunk 5]
```diff
@@ -192,7 +195,12 @@ impl MergeScanExec {
                     header: Some(RegionRequestHeader {
                         tracing_context: tracing_context.to_w3c(),
                         dbname: dbname.clone(),
-                        timezone: tz.clone(),
+                        query_context: Some(QueryContext {
+                            current_catalog: current_catalog.clone(),
+                            current_schema: current_schema.clone(),
+                            timezone: timezone.clone(),
+                            extensions: extensions.clone(),
```
Comment (killme2008) [inline src/query/src/dist_plan/merge_scan.rs:202]:
```suggestion
                            current_catalog,
                            current_schema,
                            timezone,
                            extensions,
```

**Review by killme2008 — APPROVED** [review]
Almost LGTM

File: src/query/src/dist_plan/merge_scan.rs [hunk 6]
```diff
@@ -192,7 +195,12 @@ impl MergeScanExec {
                     header: Some(RegionRequestHeader {
                         tracing_context: tracing_context.to_w3c(),
                         dbname: dbname.clone(),
-                        timezone: tz.clone(),
+                        query_context: Some(QueryContext {
+                            current_catalog,
+                            current_schema,
+                            timezone,
+                            extensions,
```
Comment (killme2008) [inline src/query/src/dist_plan/merge_scan.rs:202]:
```suggestion
                            current_catalog: current_catalog.clone(),
                            current_schema: current_schema.clone(),
                            timezone: timezone.clone(),
                            extensions: extensions.clone(),
```

File: src/query/src/dist_plan/merge_scan.rs (hunk 5 above)
Comment (Kelvinyu1117) [reply src/query/src/dist_plan/merge_scan.rs]: It seems the test is failed due to this?

Comment (killme2008) [reply src/query/src/dist_plan/merge_scan.rs:202]: Yes, revert it. It's in a loop, my mistake.

**Pull request merged**